odl/solvers/functional/functional.py

```diff
@@ -1268,6 +1333,186 @@ def __str__(self):
+
+        Parameters
+        ----------
+        functional : `Functional`
+            Functional on which to base the Bregman distance. If the
+            optional argument `subgrad` is not given, the functional
+            needs to implement `functional.gradient`.
+        point : element of ``functional.domain``
+            Point from which to define the Bregman distance.
+        subgrad : `Operator`, optional
+            A subgradient operator. Operator which that takes an element in
+            `functional.domain` and returns a subdifferential of the functional
+            in that point.
+            Default: ``functional.gradient``.
+
+        Examples
+        --------
+        Example of initializing the Bregman distance functional:
+
+        >>> space = odl.uniform_discr(0, 1, 10)
+        >>> l2_squared = odl.solvers.L2NormSquared(space)
+        >>> point = space.one()
+        >>> Bregman_dist = odl.solvers.BregmanDistance(l2_squared, point)
+
+        This is gives squared L2 distance to the given point, ||x - 1||^2:
+
+        >>> expected_functional = odl.solvers.L2NormSquared(space).translated(
+        ...     point)
+        >>> Bregman_dist(space.zero()) == expected_functional(space.zero())
+        True
+        """
+        if not isinstance(functional, Functional):
+            raise TypeError('`functional` {} not an instance of ``Functional``'
+                            ''.format(functional))
+        self.__functional = functional
+
+        if point not in functional.domain:
+            raise ValueError('`point` {} is not in `functional.domain` {}'
+                             ''.format(point, functional.domain))
+        self.__point = point
+
+        if subgrad is None:
+            # If the subgradient is not given, assign the gradient of the
+            # functional
+            try:
+                self.__subgrad = self.__functional.gradient
+            except NotImplementedError:
+                raise NotImplementedError(
+                    '`subgradient` not given, and the `functional` {} does '
+                    'not implement `functional.gradient`'.format(functional))
+        else:
+            # Check that given subgradient is an operator that maps from the
+            # domain of the functional to itself
+            if not isinstance(subgrad, Operator):
+                raise TypeError(
+                    '`subgrad` must be an instance of ``Operator``, got {}'
+                    ''.format(subgrad))
+            if subgrad.domain != self.__functional.domain:
+                raise ValueError('`functional.domain` {} is not the same as '
+                                 '`subgrad.domain` {}'
+                                 ''.format(self.__functional.domain,
+                                           subgrad.domain))
+            if subgrad.range != self.__functional.domain:
+                raise ValueError('`functional.domain` {} is not the same as '
```

Comment:
I recommend changing the order of this exception text to follow the if statement, easier to read.